### ISSUE TITLE
Policy for RuntimeWarnings for dask.dataframe

### ISSUE THREAD

**Issue by TomAugspurger** [dask/dask#2381]
(going through the pytest warnings, and splitting off issues that might be real)

Pandas silences most RuntimeWarnings from numpy when an operation involving NaN occurs.
dask.dataframe doesn't do this (at least not everywhere).

```python
In [1]: import dask.dataframe as dd
In [2]: import pandas as pd

In [3]: df = pd.Series([])

In [4]: df.std()
Out[4]: nan

In [5]: dd.from_pandas(df, npartitions=2).std().compute()
/Users/taugspurger/Envs/dask-dev/lib/python3.6/site-packages/dask/dask/dataframe/methods.py:103: RuntimeWarning: invalid value encountered in double_scalars
  result = (x2 / n) - (x / n)**2
Out[5]: nan

In [6]: df.values.std()  # numpy
/Users/taugspurger/.virtualenvs/dask-dev/lib/python3.6/site-packages/numpy/core/_methods.py:135: RuntimeWarning: Degrees of freedom <= 0 for slice
  keepdims=keepdims)
/Users/taugspurger/.virtualenvs/dask-dev/lib/python3.6/site-packages/numpy/core/_methods.py:105: RuntimeWarning: invalid value encountered in true_divide
  arrmean, rcount, out=arrmean, casting='unsafe', subok=False)
/Users/taugspurger/.virtualenvs/dask-dev/lib/python3.6/site-packages/numpy/core/_methods.py:127: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
Out[6]: nan
```

Are there any reasons dask.dataframe shouldn't follow pandas here? Potentially some issues around empty partitions that might warrant a warning?

**Comment by jcrist** [comment]
This behavior isn't intentional. However it is a bit tricky - since `numpy.seterr` is thread-local, you can't just call `seterr` in the main thread and have it take effect in other threads (or processes). Does pandas wrap calls in `numpy.errstate`? Perhaps we should do that for known offenders (or rewrite to not warn)?

**Comment by TomAugspurger** [comment]
> Does pandas wrap calls in numpy.errstate? Perhaps we should do that for known offenders (or rewrite to not warn)?

Yes, pandas does (or at least intends to, some slip through: https://github.com/pandas-dev/pandas/issues/16306)

For things like https://github.com/dask/dask/blob/master/dask/dataframe/methods.py#L101 we should be able to wrap them in `np.errstate` calls. If we want to follow pandas here I can see how difficult it is.

**Comment by jcrist** [comment]
> If we want to follow pandas here I can see how difficult it is.

Sounds good to me.

**Comment by mrocklin** [comment]
We also get many warnings in dask.array just from the small computations we perform to check metadata.  These all happen in the main thread.

```python
In [1]: import dask.array as da

In [2]: x2 = da.random.uniform(0.01,0.1,size=1e6,chunks= 1e5)
   ...: asd2 = da.arccos(x2)
   ...: 
/home/mrocklin/workspace/dask/dask/array/core.py:476: RuntimeWarning: invalid value encountered in arccos
  o = func(*args, **kwargs)
```

User reported in https://stackoverflow.com/questions/44285196/dask-python-invalid-value-encountered-in-arccos

Do we want to call `np.seterr('ignore')` when the user imports dask.array?  This would hide these warnings but would also affect a numpy global?  We could also construct a context manager.

**Comment by TomAugspurger** [comment]
> Do we want to call np.seterr('ignore') when the user imports dask.array? 

I don't think so. pandas used to do that, but the global nature of `seterr` can hide actual warnings a user would like to see. Robert Kern wen through a lot of work in https://github.com/pandas-dev/pandas/pull/13145 to use the `np.errstate` context manager only where required. That's probable what dask should do as well.
